presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFiltersCollector.java

```diff
@@ -13,40 +13,155 @@
-     */
-    @GuardedBy("this")
-    private final Map<Symbol, Domain> dynamicFilterDomainsResult = new HashMap<>();
-
-    public synchronized TupleDomain<Symbol> getDynamicFilter(Set<Symbol> probeSymbols)
+    // Each future blocks until its dynamic filter is collected.
+    private final Map<DynamicFilterId, SettableFuture<Domain>> futures = new HashMap<>();
+
+    public LocalDynamicFiltersCollector()
+    {
+    }
+
+    // Called during JoinNode planning (no need to be synchronized as local planning is single threaded)
+    public void register(Set<DynamicFilterId> filterIds)
+    {
+        filterIds.forEach(filterId -> verify(
+                futures.put(filterId, SettableFuture.create()) == null,
+                "LocalDynamicFiltersCollector: duplicate filter %s", filterId));
+    }
+
+    // Used during execution (after build-side dynamic filter collection is over).
+    // No need to be synchronized as the futures map doesn't change.
+    public void collectDynamicTupleDomain(Map<DynamicFilterId, Domain> dynamicFilterDomains)
+    {
+        dynamicFilterDomains
+                .entrySet()
+                .stream()
+                .forEach(entry -> {
+                    SettableFuture<Domain> future = futures.get(entry.getKey());
+                    // Skip dynamic filters that are not applied locally.
+                    if (future != null) {
+                        verify(future.set(entry.getValue()), "Dynamic filter %s already collected", entry.getKey());
+                    }
+                });
+    }
+
+    // Called during TableScan planning (no need to be synchronized as local planning is single threaded)
+    public DynamicFilter createDynamicFilter(List<DynamicFilters.Descriptor> descriptors, Map<Symbol, ColumnHandle> columnsMap)
     {
-        Map<Symbol, Domain> probeSymbolDomains = dynamicFilterDomainsResult.entrySet().stream()
-                .filter(entry -> probeSymbols.contains(entry.getKey()))
-                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
-        return TupleDomain.withColumnDomains(probeSymbolDomains);
+        ImmutableMultimap.Builder<DynamicFilterId, Symbol> symbolsMapBuilder = ImmutableMultimap.builder();
+        for (DynamicFilters.Descriptor descriptor : descriptors) {
+            symbolsMapBuilder.put(descriptor.getId(), Symbol.from(descriptor.getInput()));
+        }
+        Multimap<DynamicFilterId, Symbol> symbolsMap = symbolsMapBuilder.build();
+
+        // Iterate over dynamic filters that are collected (correspond to one of the futures), and required for filtering (correspond to one of the descriptors).
+        // It is possible that some dynamic filters are collected in a different stage - and will not available here.
+        // It is also possible that not all local dynamic filters are needed for this specific table scan.
+        Set<DynamicFilterId> filterIds = Sets.intersection(symbolsMap.keySet(), futures.keySet());
```

Comment:
I am not sure whether it is correct to replace:
```java
Set<DynamicFilterId> filterIds = Sets.intersection(symbolsMap.keySet(), futures.keySet());
```
with:
```java
Set<DynamicFilterId> filterIds = symbolsMap.keySet();
```
For example, in case of a partitioned join, `symbolsMap.keySet()` may contain dynamic filters collected in another stage - causing the following line to fail:
```java
requireNonNull(futures.get(filterId), () -> format("Missing dynamic filter %s", filterId))
```

Reply:
You are right. We cannot simplify it.